commit 698e5f304f4b
Author: Gijs Kruitbosch <gijskruitbosch@gmail.com>
Date:   Wed Oct 4 13:29:53 2017 +0100

    Bug 1404651 - Keep focus in the tabstrip when the focus has shifted with the arrow keys. r=dao, a=ritu
    
    MozReview-Commit-ID: Brm9jImW9bc
    
    --HG--
    extra : source : 0304a90aee90618165af50193ea6255c54a6b9ea
---
 browser/base/content/tabbrowser.xml | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git browser/base/content/tabbrowser.xml browser/base/content/tabbrowser.xml
index 37efdcc5bc0c..4f40e45f2446 100644
--- browser/base/content/tabbrowser.xml
+++ browser/base/content/tabbrowser.xml
@@ -1417,13 +1417,14 @@
               findBar._findField.getAttribute("focused") == "true");
           }
 
-          // If focus is in the tab bar, retain it there.
-          if (document.activeElement == oldTab) {
-            // We need to explicitly focus the new tab, because
-            // tabbox.xml does this only in some cases.
+          let activeEl = document.activeElement;
+          // If focus is on the old tab, move it to the new tab.
+          if (activeEl == oldTab) {
             newTab.focus();
-          } else if (gMultiProcessBrowser && document.activeElement !== newBrowser) {
-
+          } else if (gMultiProcessBrowser && activeEl != newBrowser && activeEl != newTab) {
+            // In e10s, if focus isn't already in the tabstrip or on the new browser,
+            // and the new browser's previous focus wasn't in the url bar but focus is
+            // there now, we need to adjust focus further.
             let keepFocusOnUrlBar = newBrowser &&
                                     newBrowser._urlbarFocused &&
                                     gURLBar &&
